src/parser/core/modules/CooldownDowntime.tsx

```diff
@@ -0,0 +1,264 @@
+	protected defaultFirstUseOffset = 0
+
+	/**
+	 * Implementing modules MAY filter out some usages as 'fake' usages
+	 * of a cooldown.
+	 *
+	 * The primary example of this would be NIN mudras, where any mudra
+	 * may start a usage, but further mudras while the Mudra buff or any
+	 * during the Kassatsu buff should not be counted as usages, even
+	 * though all 3 mudra skills are part of the group.
+	 *
+	 * @param event The cast that may or may not be counted.
+	 * @returns True if the event should be counted or false if the event
+	 * should not be counted as a usage of the cooldown.
+	 */
+	protected countUsage(event: CastEvent): boolean {
+		return true
+	}
+
+	protected init() {
+		const trackedIds = this.trackedCds.map(g => g.cooldowns)
+			.reduce((acc, cur) => acc.concat(cur))
+			.map(a => a.id)
+
+		const resetIds = this.trackedCds
+			.map(g => (g.resetBy === undefined) ? [] : g.resetBy.actions)
+			.reduce((acc, cur) => acc.concat(cur))
+			.map(a => a.id)
+
+		this.addEventHook('cast', {by: 'player', abilityId: trackedIds}, this.onTrackedCast)
+		this.addEventHook('cast', {by: 'player', abilityId: resetIds}, this.onResetCast)
+		this.addEventHook('complete', this.onComplete)
+
+		this.trackedCds.forEach(g => {
+			this.usages.set(g.cooldowns[0].id, [])
+			this.resets.set(g.cooldowns[0].id, [])
+		})
+	}
+
+	private onTrackedCast(event: CastEvent) {
+		if (!this.countUsage(event)) { return }
+
+		const groupId = this.getTrackedGroupId(event.ability.guid)
+		if (groupId === undefined) { return }
+
+		(this.usages.get(groupId) || []).push(event)
+	}
+
+	private getTrackedGroupId(abilityId: number): number | undefined {
+		const group = this.trackedCds.find(g => g.cooldowns.find(a => a.id === abilityId) !== undefined)
+		if (group === undefined) { return undefined }
+		return group.cooldowns[0].id
+	}
+
+	private onResetCast(event: CastEvent) {
+		this.trackedCds.forEach(g => {
+			if (g.resetBy && g.resetBy.actions.find(a => a.id === event.ability.guid)) {
+				(this.resets.get(g.cooldowns[0].id) || [] ).push(event)
+			}
+		})
+	}
+
+	private onComplete() {
+		const cdRequirements = []
+		for (const cdGroup of this.trackedCds) {
+			const expected = this.calculateMaxUsages(cdGroup)
+			const actual = (this.usages.get(cdGroup.cooldowns[0].id) || []).length || 0
+			const percent = actual / expected * 100
+			const requirementDisplay = cdGroup.cooldowns.map((val, ix) => <>{( ix > 0 ? ', ' : '' )}<ActionLink {...getDataBy(ACTIONS, 'id', val.id)} /></>)
+			this.debugLog(JSON.stringify(requirementDisplay))
+
+			cdRequirements.push(new Requirement({
+				name: requirementDisplay,
+				percent,
+				overrideDisplay: `${actual} / ${expected} (${percent.toFixed(2)}%)`,
+			}))
+		}
+
+		this.checklist.add(new Rule({
+			name: this.checklistName,
+			description: this.checklistDescription,
+			requirements: cdRequirements,
+			target: this.checklistTarget,
+		}))
+	}
+
+	private calculateMaxUsages(group: CooldownGroup): number {
```

Comment:
aaaaaaaaaaaaaaaaaaaaaaaaaaaaite my brain shut down when i got here, too late

---

note to akk, check the rest of the PR after this message